src/lib/gadgets/elliptic-curve.ts

```diff
@@ -0,0 +1,575 @@
+  ForeignField.assertLessThan(Rx, Curve.order);
+
+  return Provable.equal(Field3.provable, Rx, r);
+}
+
+/**
+ * Multi-scalar multiplication:
+ *
+ * s_0 * P_0 + ... + s_(n-1) * P_(n-1)
+ *
+ * where P_i are any points. The result is not allowed to be zero.
+ *
+ * We double all points together and leverage a precomputed table of size 2^c to avoid all but every cth addition.
+ *
+ * Note: this algorithm targets a small number of points, like 2 needed for ECDSA verification.
+ *
+ * TODO: could use lookups for picking precomputed multiples, instead of O(2^c) provable switch
+ * TODO: custom bit representation for the scalar that avoids 0, to get rid of the degenerate addition case
+ * TODO: glv trick which cuts down ec doubles by half by splitting s*P = s0*P + s1*endo(P) with s0, s1 in [0, 2^128)
+ */
+function multiScalarMul(
+  Curve: CurveAffine,
+  scalars: Field3[],
+  points: Point[],
+  tableConfigs: (
+    | { windowSize?: number; multiples?: Point[] }
+    | undefined
+  )[] = [],
+  ia?: point
+): Point {
+  let n = points.length;
+  assert(scalars.length === n, 'Points and scalars lengths must match');
+  assertPositiveInteger(n, 'Expected at least 1 point and scalar');
+
+  // constant case
+  if (scalars.every(Field3.isConstant) && points.every(Point.isConstant)) {
+    // TODO dedicated MSM
+    let s = scalars.map(Field3.toBigint);
+    let P = points.map(Point.toBigint);
+    let sum = Curve.zero;
+    for (let i = 0; i < n; i++) {
+      sum = Curve.add(sum, Curve.scale(P[i], s[i]));
+    }
+    return Point.from(sum);
+  }
+
+  // parse or build point tables
+  let windowSizes = points.map((_, i) => tableConfigs[i]?.windowSize ?? 1);
+  let tables = points.map((P, i) =>
+    getPointTable(Curve, P, windowSizes[i], tableConfigs[i]?.multiples)
+  );
+
+  // slice scalars
+  let b = Curve.order.toString(2).length;
+  let scalarChunks = scalars.map((s, i) =>
+    slice(s, { maxBits: b, chunkSize: windowSizes[i] })
+  );
+
+  ia ??= initialAggregator(Curve);
+  let sum = Point.from(ia);
+
+  for (let i = b - 1; i >= 0; i--) {
+    // add in multiple of each point
+    for (let j = 0; j < n; j++) {
+      let windowSize = windowSizes[j];
+      if (i % windowSize === 0) {
+        // pick point to add based on the scalar chunk
+        let sj = scalarChunks[j][i / windowSize];
+        let sjP =
+          windowSize === 1
+            ? points[j]
+            : arrayGetGeneric(Point.provable, tables[j], sj);
+
+        // ec addition
+        let added = add(sum, sjP, Curve.modulus);
+
+        // handle degenerate case (if sj = 0, Gj is all zeros and the add result is garbage)
+        sum = Provable.if(sj.equals(0), Point.provable, sum, added);
+      }
+    }
+
+    if (i === 0) break;
+
+    // jointly double all points
+    // (note: the highest couple of bits will not create any constraints because sum is constant; no need to handle that explicitly)
```

Comment:
so we start the sum with just the `ia`. and we use a window size of, say, 3 and 4. then, in the first 3 iterations of the loop, neither G nor P are added yet. So nothing will happen except this doubling of the `sum`. Since the `sum` was constant in the beginning, this will just double a constant, and not create constraints. Only when the first variable is added (in the form of either point) will the first constraint be created. At first I though I should start the loop at `maxBits - 3` manually to avoid those first 3 iterations, but then realized that they don't cost anything anyway